config/src/lib.rs

```diff
@@ -0,0 +1,482 @@
+//! foundry configuration.
+use ethers_core::types::Address;
+use ethers_solc::{remappings::Remapping, ProjectPathsConfig};
+use figment::{
+    providers::{Env, Format, Serialized, Toml},
+    value::{Dict, Map},
+    Figment, Metadata, Profile, Provider,
+};
+use semver::Version;
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, path::PathBuf};
+
+/// Foundry configuration
+#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
+pub struct Config {
+    /// The selected profile. **(default: _default_ `default`)**
+    ///
+    /// **Note:** This field is never serialized nor deserialized. When a
+    /// `Config` is merged into a `Figment` as a `Provider`, this profile is
+    /// selected on the `Figment`. When a `Config` is extracted, this field is
+    /// set to the extracting Figment's selected `Profile`.
+    #[serde(skip)]
+    pub profile: Profile,
+    /// path of the source contracts dir, like `src` or `contracts`
+    pub src: PathBuf,
+    /// path of the test dir
+    pub test: PathBuf,
+    /// path to where artifacts shut be written to
+    pub out: PathBuf,
+    /// all library folders to include, `lib`, `node_modules`
+    pub libs: Vec<PathBuf>,
+    /// whether to enable cache
+    pub cache: bool,
+    /// concrete solc version to use if any,
+    pub solc_version: Option<Version>,
+    /// Whether to activate optimizer
+    pub optimizer: bool,
+    /// Sets the optimizer runs
+    pub optimizer_runs: usize,
+    /// Settings to pass to the `solc` compiler input
+    // TODO make this more flexible https://stackoverflow.com/questions/48998034/does-toml-support-nested-arrays-of-objects-tables
+    pub solc_settings: serde_json::Value,
+    /// url of the rpc server that should be used for any rpc calls
+    pub eth_rpc_url: Option<String>,
+    /// verbosity to use
+    pub verbosity: u8,
+    /// `Remappings` to use for this repo
+    pub remappings: Vec<Remapping>,
+    /// library addresses to link
+    pub libraries: Vec<Address>,
+
+    /// PRIVATE: This structure may grow, As such, constructing this structure should
+    /// _always_ be done using a public constructor or update syntax:
```

Comment:
Hm typically you'd do this with all fields private, and only exposing getters, maybe worth using something like https://crates.io/crates/derive-getters?

Reply:
this enforces that we always use this pattern with the trailing `..Config::default()`, so that, if we add new config vars we don't end up with `missing field` values

```rust
Config {
    src: "contracts".into(),
    out: "artifacts".into(),
    libs: vec!["node_modules".into()],
    ..Config::default()
}
```